### PR TITLE
tighten CTFE safety net for accesses to globals

### PR DESCRIPTION
**Pull request by RalfJung**
Previously, we only rejected reading from all statics. Now we also reject reading from any mutable global. Mutable globals are the true culprit here as their run-time value might be different from their compile-time values. Statics are just the approximation we use for that so far.

Also refactor the code a bit to make it clearer what is being checked and allowed.

r? @oli-obk

### CODE REVIEW

File: src/librustc_mir/const_eval/machine.rs [hunk 1]
```diff
@@ -350,15 +350,30 @@ impl<'mir, 'tcx> interpret::Machine<'mir, 'tcx> for CompileTimeInterpreter {
...
+            } else {
+                Err(ConstEvalErrKind::ModifiedGlobal.into())
+            }
         } else {
-            Err(ConstEvalErrKind::ConstAccessesStatic.into())
+            // Read access. These are usually allowed, with some exceptions.
+            if memory_extra.can_access_statics {
+                // This is allowed to read from anything.
+                Ok(())
+            } else if allocation.mutability == Mutability::Mut || static_def_id.is_some() {
```
Comment (oli-obk) [inline src/librustc_mir/const_eval/machine.rs:365]: At present, shouldn't `allocation.mutability` be false for all non-statics? If so, can we assert it in the `else` branch?

File: src/librustc_mir/const_eval/machine.rs [hunk 2]
```diff
@@ -350,15 +350,30 @@ impl<'mir, 'tcx> interpret::Machine<'mir, 'tcx> for CompileTimeInterpreter {
...
+            if allocation.mutability == Mutability::Not {
+                Err(err_ub!(WriteToReadOnly(alloc_id)).into())
+            } else {
+                Err(ConstEvalErrKind::ModifiedGlobal.into())
+            }
         } else {
-            Err(ConstEvalErrKind::ConstAccessesStatic.into())
+            // Read access. These are usually allowed, with some exceptions.
+            if memory_extra.can_access_statics {
+                // This is allowed to read from anything.
```
Comment (Centril) [inline src/librustc_mir/const_eval/machine.rs:363]: Could you clarify "this" here -- perhaps with an example (and below)?

Comment (RalfJung) [reply src/librustc_mir/const_eval/machine.rs:363]: I can clarify to "this machine" or "this const evaluation", but I am not sure what kind of "example" you are asking for.

Comment (RalfJung) [reply src/librustc_mir/const_eval/machine.rs]: I did expand the comments a bit, does this help @Centril?

File: src/librustc_mir/const_eval/machine.rs (hunk 1 above)
Comment (RalfJung) [reply src/librustc_mir/const_eval/machine.rs]: Looks like you are right. I thought some promoteds would still be mutable but forgot that we make them [immutable during interning](https://github.com/rust-lang/rust/blob/0c835b0cca83fe21090562603e4bda77c183ace3/src/librustc_mir/interpret/intern.rs#L369).

File: src/librustc_mir/const_eval/machine.rs (hunk 2 above)
Comment (Centril) [reply src/librustc_mir/const_eval/machine.rs]: It does, thank you Ralf.

**Pull request merged**